src/MongoDB.Driver/Linq/Translators/PredicateTranslator.cs

```diff
@@ -217,6 +217,32 @@ private bool CanAnyBeRenderedWithoutElemMatch(Expression node)
             }
         }
 
+        private FilterDefinition<BsonDocument> ConvertElemMatchFilterToScalarElementMatchIfNeeded(FilterDefinition<BsonDocument> filter, IFieldExpression fieldExpression, Expression wherePredicate)
+        {
+            if ((!(fieldExpression.Serializer is IBsonDocumentSerializer)) || DoesExpressionUseDocumentItself(wherePredicate))
```

Comment:
The previous code is tricky a little bit:

    (!(fieldExpression.Serializer is IBsonDocumentSerializer)) 

It looks like the problem of this ticket happened because of user used the following collection:

    `public ICollection<ObjectId> ParentIds { get; set; }`. 

The serializer for this field(`ICollection`) inherits `IBsonDocumentSerializer` and the condition doesn't work correctly. 
In case if user uses something like:

    public ObjectId[] ParentIds { get; set; }`

he would not face with this error since an array do not inherit `IBsonDocumentSerializer`.  And condition will work properly ('ScalarElementMatchFilterDefinition' will be called).  

@craiggwilson can you help here, could you please take a look on this comment? I'm not sure how this logic was expected to use?

Reply:
So, the fact that ICollection<ObjectId> ParentIds is using an IBsonDocumentSerializer is problematic. This would ideally get rendered to an array, but they had this misconfigured (or not configured at all). Not sure if that answers your question, but I would solve this by having them fix their mapping rather than us fixing this in code.

Reply:
Yep, I also saw that `array` data is processed in a different way. But not sure that we can require this restriction from users

Reply:
Dima and I discussed. He's going to try another method where we don't need to check on the type of serializer we are using. I think it might cause some issues because certain serializers (ImpliedInterfaceImplementationSerializer) implements the interface, but it's a lie!

Reply:
I'm not sure I agree it's a lie. It actually does correctly implement the interface.

Reply:
Agree... it does correctly implement it. But checking whether a serializer implements the interface doesn't tell you if it will serialize as a document or an array. Hence, checking for the presence of an interface can't be trusted.

Reply:
I'm still wondering why we need to IBsonDocumentSerializer check. If I understand the reasoning here, this check doesn't matter. I know we discussed this... Did we test this hypothesis?

Reply:
>Did we test this hypothesis?
When I removed the checking on `IBsonDocumentSerializer`, almost all test passed, except of the one:

        Assert(
                x => x.M.Any(m => m > 2 && m < 6),
                2,
                "{ M : { $elemMatch : { $gt : 2, $lt : 6 } } }");

This test didn't pass since the `Left` expression in `BinaryExpression` was `LogicalBinaryExpression` and not `DocumentExpression` itself.
I added several commented out todo tests with a similar idea. I guess it should be fixed in a separate ticket.
So, it looks like the tests didn't have the difference, but I still worry that there are cases which we don't consider. 
So, my unser here:
>I'm still wondering why we need to IBsonDocumentSerializer check.
to have the solution safer.

Reply:
>to have the solution safer.

in case if my solution doesn't work for some cases, we will still have the previous solution. And at least we will not have the problem as the user has in this ticket. When we've changed the logic which wasn't connected with the user problem directly, but since it was common logic, the change also affected the user. 

Maybe we can delete checking on the serializer interface in the ticket which will solve the `todo` tests.

Reply:
There is also a chance that having the check here breaks something. For instance, you are basing this on the fact that ImpliedInterfaceImplementationSerializer should always be an array because it is for ICollection. But what if it is _supposed_ to be a document?

Reply:
Do you mean the query similar to this: 

    x => x.Document.Any(g => g != null),

where `Document` is `BsonDocument`?

if so, currently this query is not supported. See [BindNonMethodCall ](https://github.com/mongodb/mongo-csharp-driver/blob/14e046f23640ff9257c4edf53065b9a6768254d4/src/MongoDB.Driver/Linq/Processors/EmbeddedPipeline/EmbeddedPipelineBinder.cs#L97)for `x.Document`. For non array serializer or non constant node typs we will have `NotSupportedException`.

Reply:
Anyway, even if this check gives us true:

    (fieldExpression.Serializer is IBsonDocumentSerializer)

then, since this checking is under the negative operand (`!`), we will must to check the second part of `or` condition. So, as I see it can not be a problem.

As I see, the only case which can be a problem, if this checking:

    (!(fieldExpression.Serializer is IBsonDocumentSerializer))

gives us an unexpected `true` result. But, I guess, it should be another ticket, since this code is already in release long time and after quick search in jira I didn't find mentions of similar issues.

Reply:
The unexpected `true` result is exactly the thing I'm concerned about.

Reply:
So, my main concern here, that the scope of this ticket should be first of all about fixing the new bug which appeared after one from my previous tickets and without creating new issues.

Maybe we can worry about this possible issue in the scope of the new `todo` ticket? @jyemin what do you think?

Reply:
Generally, assuming we can get consensus, I'm not opposed to a targeted fix that addresses the bug and doesn't introduce any new regressions.

Reply:
Y'all's call. It doesn't seem right, but I've been away from this code a long time. If this really is targeted and won't cause a different regression (my main concern), then fine by me.

Reply:
>won't cause a different regression (my main concern)

since it's the legacy [code](https://github.com/mongodb/mongo-csharp-driver/blob/efeda17a0871b3a068061d45f7f8f472777814b5/src/MongoDB.Driver/Linq/Translators/PredicateTranslator.cs#L938), I don't see how it can cause different regression.

Reply:
Why do you say it's legacy code?

Reply:
I don't see how we can base any decisions on whether a serializer implements IBsonDocumentSerializer or not.

Since implementing IBsonDocumentSerializer is optional just because a serializer doesn't implement it doesn't mean that the value is not a document.

And even if a serializer does implement IBsonDocumentSerializer, that also doesn't mean that value is a document (e.g. ImpliedImplementationInterfaceSerializer).

And even if none of the above weirdnesses occur and the value is in fact a document, we still might end up generating an $elemMatch with an empty string for the field name. Doesn't it depend on whether we are querying on a field of a document, not whether the value is a document or not?

Reply:
>Why do you say it's legacy code?
I meant that the below comparison:

    (!(fieldExpression.Serializer is IBsonDocumentSerializer))

is the legacy code.
It was the answer on the Craig's queries:
>I'm still wondering why we need to IBsonDocumentSerializer check. 

>There is also a chance that having the check here breaks something.
and on his last comment:

>The unexpected true result is exactly the thing I'm concerned about.

so, as we discussed with him, the legacy checking on serializer type is problematic and can give some unexpected problem for example the issue which we see in this ticket. In theory `DoesExpressionUseDocumentItself` covers all cases here, and we don't need checking on `IBsonDocumentSerializer` anymore. But I prefer to leave this checking as It is at least for now, just to prevent generation wrong queries for cases which `DoesExpressionUseDocumentItself` possibly don't consider for now. 

We can remove this checking in a new "todo" ticket, maybe after implementation more complex ticket, the possible issues will be clearer.

I see, the only new problem which we can introduce here is wrong logic in: `DoesExpressionUseDocumentItself`

Reply:
> I don't see how we can base any decisions on whether a serializer implements IBsonDocumentSerializer or not.

>Since implementing IBsonDocumentSerializer is optional just because a serializer doesn't implement it doesn't mean that the value is not a document.

>And even if a serializer does implement IBsonDocumentSerializer, that also doesn't mean that value is a document (e.g. ImpliedImplementationInterfaceSerializer).

Agree. But since it's the legacy code and I didn't find after a quick search in Jira any tickets about this logic, currently, I don't want to remove this logic, exactly for reason that I'm not sure how this change can affect query generating for different cases. So, *to me* having legacy checking together with `DoesExpressionUseDocumentItself` just safer for now.

Reply:
Usually "legacy" means something really old, like what is in the MongoDB.Driver.Legacy project.

So I'm still not sure what you mean by "legacy". How far back in time are you going?

Reply:
>And even if none of the above weirdnesses occur and the value is in fact a document, we still might end up generating an $elemMatch with an empty string for the field name. Doesn't it depend on whether we are querying on a field of a document, not whether the value is a document or not?

Not sure that I understood your question correctly.
As I see, it depends on whether we have the `.` part in expression operands or not.
Let's consider two examples:

>Any(p=> p.Name == "test")

as I see, this one will not have gaps in the generated query. The expression operand is `FieldExpresion`

>Any(p=> p == "test")

but this one will have gaps since p is `DocumentExpression`.

Reply:
I'm not sure I understand my question correctly either... :)

I was just questioning whether the type of the serializer is the right to base decisions on.

Reply:
>So I'm still not sure what you mean by "legacy". How far back in time are you going?

I meant the code before changes which I applied in the ticket 2012.  Maybe, the `legacy` word makes sense here only for me :), since I consider all previously written code before I came to project as `legacy`. But you are right, it's not very clear.
Let's use `the code which looks correct and doesn't have JIRA ticket` :)

>I was just questioning whether the type of the serializer is the right to base decisions on.

The answer is it's wrong (at least unreliable), but I'm not 100% sure that we can remove this part without any consequences.